File: reposteria_flutter/lib/models/producto.dart
import 'producto_variante.dart';
import 'promocion_precio.dart';

class Producto {
  const Producto({
    required this.id,
    required this.categoriaId,
    required this.nombre,
    required this.descripcion,
    required this.precio,
    required this.precioFinal,
    required this.imagen,
    required this.personalizable,
    required this.manejaStock,
    required this.stock,
    required this.tieneVariantes,
    required this.promocion,
    required this.variantes,
  });

  final int id;
  final int? categoriaId;
  final String nombre;
  final String? descripcion;
  final String precio;
  final String precioFinal;
  final String? imagen;
  final bool personalizable;
  final bool manejaStock;
  final int stock;
  final bool tieneVariantes;
  final PromocionPrecio? promocion;
  final List<ProductoVariante> variantes;

  bool get disponible => !manejaStock || stock > 0 || tieneVariantes;

  factory Producto.fromJson(Map<String, dynamic> json) => Producto(
    id: json['id'] as int,
    categoriaId: json['categoria_id'] as int?,
    nombre: json['nombre'] as String,
    descripcion: json['descripcion'] as String?,
    precio: json['precio'] as String,
    precioFinal: json['precio_final'] as String,
    imagen: json['imagen'] as String?,
    personalizable: json['personalizable'] as bool,
    manejaStock: json['maneja_stock'] as bool,
    stock: json['stock'] as int,
    tieneVariantes: json['tiene_variantes'] as bool,
    promocion: json['promocion'] == null
        ? null
        : PromocionPrecio.fromJson(json['promocion'] as Map<String, dynamic>),
    variantes: (json['variantes'] as List<dynamic>? ?? const [])
        .map((item) => ProductoVariante.fromJson(item as Map<String, dynamic>))
        .toList(growable: false),
  );
}

File: reposteria-api/app/Models/Producto.php
<?php

namespace App\Models;

use Database\Factories\ProductoFactory;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Producto extends Model
{
    /** @use HasFactory<ProductoFactory> */
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'nombre', 'descripcion', 'precio', 'imagen', 'personalizable', 'maneja_stock', 'stock',
    ];

    public function reposteria(): BelongsTo
    {
        return $this->belongsTo(Reposteria::class);
    }

    public function categoria(): BelongsTo
    {
        return $this->belongsTo(Categoria::class);
    }

    public function variantes(): HasMany
    {
        return $this->hasMany(ProductoVariante::class);
    }

    protected function casts(): array
    {
        return [
            'precio' => 'decimal:2',
            'personalizable' => 'boolean',
            'maneja_stock' => 'boolean',
            'stock' => 'integer',
            'activo' => 'boolean',
            'deleted_at' => 'datetime',
        ];
    }
}
